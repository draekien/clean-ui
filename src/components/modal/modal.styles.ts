import { ThemeUIStyleObject } from '@theme-ui/css';

export const modalContainerCss: ThemeUIStyleObject = {
  zIndex: 'modal',
};

export const modalWrapperCss: ThemeUIStyleObject = {
  opacity: 0,
  transition: 'all 300ms ease-in-out',
  zIndex: 'modal',
};

export const modalWrapperTransitionCss: any = {
  entering: { opacity: 0 },
  entered: { opacity: 1 },
  exiting: { opacity: 0 },
  exited: { opacity: 0 },
};
