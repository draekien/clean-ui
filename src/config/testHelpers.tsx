import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
