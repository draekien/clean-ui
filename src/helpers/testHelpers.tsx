import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
