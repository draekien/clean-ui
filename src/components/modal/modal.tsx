/** @jsxImportSource theme-ui */
import React from 'react';
import { Overlay } from '../overlay/overlay';
import { Card } from '../card/card';
import { Button } from '../button/button';
import { modalWrapperCss, modalWrapperTransitionCss } from './modal.styles';
import OutsideClickHandler from 'react-outside-click-handler';
import { Transition } from 'react-transition-group';

export interface ModalProps {
  header?: React.ReactNode;
  footer?: React.ReactNode;
  width?: string | number | string[] | number[];
  visible?: boolean;
  feature?: boolean;
  fullPage?: boolean;
  noPadding?: boolean;
  showCloseButton?: boolean;
  onClose?: () => void;
}

export const Modal: React.FC<ModalProps> = ({
  fullPage = false,
  children,
  onClose,
  showCloseButton = false,
  header,
  visible = false,
  ...rest
}) => {
  const handleOutsideClicked = () => {
    onClose && onClose();
  };

  const closeButton = showCloseButton && (
    <Button
      variant="text"
      icon="close"
      iconPosition="right"
      onClick={handleOutsideClicked}
    />
  );

  const modalHeader = (
    <div
      sx={{
        display: 'flex',
        flexFlow: 'row nowrap',
        justifyContent: 'space-between',
        alignItems: 'flex-start',
      }}>
      {header}
      {closeButton}
    </div>
  );

  const card = (
    <OutsideClickHandler onOutsideClick={handleOutsideClicked}>
      <Card {...rest} variant="large" header={modalHeader}>
        {children}
      </Card>
    </OutsideClickHandler>
  );

  const cardWithOverlay = <Overlay>{card}</Overlay>;

  return (
    <Transition appear in={visible} timeout={0}>
      {(state) => (
        <div sx={{ ...modalWrapperCss, ...modalWrapperTransitionCss[state] }}>
          {fullPage ? cardWithOverlay : card}
        </div>
      )}
    </Transition>
  );
};
