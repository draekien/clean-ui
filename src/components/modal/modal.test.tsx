import React from 'react';
import { themedRender } from '../../helpers/testHelpers';
import { Modal } from './modal';

describe('Modal component', () => {
  it('should match snapshot', () => {
    const { container } = themedRender(<Modal>test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when fullPage is true', () => {
    const { container } = themedRender(<Modal fullPage>test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when visible is true', () => {
    const { container } = themedRender(<Modal visible>test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when showCloseButton is true', () => {
    const { container } = themedRender(<Modal showCloseButton>test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when feature is true', () => {
    const { container } = themedRender(<Modal feature>test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when noPadding is true', () => {
    const { container } = themedRender(<Modal noPadding>test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when width is set', () => {
    const { container } = themedRender(<Modal width="40em">test</Modal>);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with header and footer', () => {
    const { container } = themedRender(
      <Modal header="header" footer="footer">
        test
      </Modal>
    );

    expect(container).toMatchSnapshot();
  });
});
