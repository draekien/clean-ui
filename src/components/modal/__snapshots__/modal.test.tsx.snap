// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should match snapshot 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-glvcy5-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              />
            </div>
            <div>
              test
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot when feature is true 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-1difpm5-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              />
            </div>
            <div>
              test
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot when fullPage is true 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div
      class="css-1hdj22u-Overlay"
    >
      <div>
        <div
          class="css-glvcy5-Card"
        >
          <div
            style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
          >
            <div
              style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
            >
              <div
                class="css-u1vxhg-Card"
              >
                <div
                  class="css-accpgg-Modal"
                />
              </div>
              <div>
                test
              </div>
            </div>
            <div
              style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
            >
              <div
                style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
              />
            </div>
            <div
              style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
            >
              <div
                style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot when noPadding is true 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-blf7or-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              />
            </div>
            <div>
              test
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot when showCloseButton is true 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-glvcy5-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              >
                <button
                  class="css-1pznfqy-Button"
                  icon="close"
                >
                  <span
                    class="css-zohjwb-ButtonIcon"
                  >
                    <i
                      class="material-icons  css-y0j37o-Icon"
                    >
                      close
                    </i>
                  </span>
                </button>
              </div>
            </div>
            <div>
              test
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot when visible is true 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-glvcy5-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              />
            </div>
            <div>
              test
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot when width is set 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-16cxr1x-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              />
            </div>
            <div>
              test
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should match snapshot with header and footer 1`] = `
<div>
  <div
    class="css-1ra6pze"
  >
    <div>
      <div
        class="css-glvcy5-Card"
      >
        <div
          style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
        >
          <div
            style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
          >
            <div
              class="css-u1vxhg-Card"
            >
              <div
                class="css-accpgg-Modal"
              >
                header
              </div>
            </div>
            <div>
              test
            </div>
            <div
              class="css-1tdomna-Card"
            >
              footer
            </div>
          </div>
          <div
            style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
          <div
            style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
          >
            <div
              style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
